src/pfs/pfs.go

```diff
@@ -184,3 +186,141 @@ func (r *Repo) AuthResource() *auth.Resource {
+	p.Picker = &ProjectPicker_Name{
+		Name: string(b),
+	}
+	if err := (&Project{Name: string(b)}).ValidateName(); err != nil {
+		return err
+	}
+	if err := p.ValidateAll(); err != nil {
+		return err
+	}
+	return nil
+}
+
+var _ encoding.TextUnmarshaler = (*ProjectPicker)(nil)
+
+type repoName Repo // repoName is a repo object used only to store the name and type of a repo.
+
+func (n *repoName) UnmarshalText(b []byte) error {
+	parts := bytes.SplitN(b, []byte{'.'}, 2)
+	switch len(parts) {
+	case 0:
+		return errors.New("invalid repo name: empty")
+	case 1:
+		n.Name = string(b)
+		n.Type = UserRepoType
+	case 2:
+		n.Name = string(parts[0])
+		n.Type = string(parts[1])
+	}
+	return nil
+}
+
+var _ encoding.TextUnmarshaler = (*repoName)(nil)
+
+// UnmarshalText implements encoding.TextUnmarshaler.
+func (p *RepoPicker) UnmarshalText(b []byte) error {
+	parts := bytes.SplitN(b, []byte{'/'}, 2)
+	var repo, project []byte
+	switch len(parts) {
+	case 0:
+		return errors.New("invalid repo picker: empty")
+	case 1:
+		repo, project = parts[0], nil
+	case 2:
+		repo, project = parts[1], parts[0]
+	default:
+		return errors.New("invalid repo picker: too many slashes")
+	}
+	rnp := &RepoPicker_RepoName{}
+	p.Picker = &RepoPicker_Name{
+		Name: rnp,
+	}
+	if project != nil {
+		rnp.Project = &ProjectPicker{}
+		if err := rnp.Project.UnmarshalText(project); err != nil {
+			return errors.Wrapf(err, "unmarshal project %s", project)
+		}
+	}
+	var repoName repoName
+	if err := repoName.UnmarshalText(repo); err != nil {
+		return errors.Wrapf(err, "unmarshal repo name %s", repo)
+	}
+	rnp.Name = repoName.Name
+	rnp.Type = repoName.Type
+	if err := p.ValidateAll(); err != nil {
+		return err
+	}
+	return nil
+}
+
+var _ encoding.TextUnmarshaler = (*RepoPicker)(nil)
+
+func onlyHex(p []byte) bool {
+	for _, b := range p {
+		if (b >= 'a' && b <= 'f') || (b >= 'A' && b <= 'F') || (b >= '0' && b <= '9') {
+			continue
+		}
+		return false
+	}
+	return true
+}
+
+func (p *CommitPicker) UnmarshalText(b []byte) error {
+	parts := bytes.SplitN(b, []byte{'@'}, 2)
+	var repo, id []byte
+	switch len(parts) {
+	case 0:
+		return errors.New("invalid commit picker: empty")
+	case 1:
+		return errors.New("invalid commit picker: no @id")
+	case 2:
+		repo, id = parts[0], parts[1]
+	default:
+		return errors.New("invalid commit picker: too many @s")
+	}
+	var rp RepoPicker
+	if err := rp.UnmarshalText(repo); err != nil {
+		return errors.Wrapf(err, "unmarshal repo picker %s", repo)
+	}
+	// TODO(PFS-229): Implement the other parsers.
+	// CommitPicker_AncestorOf
+	if bytes.HasSuffix(id, []byte{'^'}) {
+		return errors.New("ancestor of commit syntax is currently unimplemented (id^); specify the exact id instead")
+	}
+	// CommitPicker_BranchRoot
+	if bytes.Contains(id, []byte{'.'}) {
+		return errors.New("branch root commit syntax is currently unimplemented (id.N); specify the exact id instead")
+	}
+	if len(id) == 32 && id[12] == '4' && onlyHex(id) {
```

Comment:
I think a short comment would be helpful here to explain to a reader that we're checking against almost commit ids.

Reply:
Done.